### PR TITLE
Split out Instrumentation from Extractor

### PR DESCRIPTION
**Pull request by chrisbutcher**
Introduces a new public method, `GraphQLMetrics::Extractor#extract!`, which takes in a `GraphQL::Query` and permits stateful extractors to be created that aren't used as instrumentation. See the test case below for an idea of its usage.

Only `field_extracted`, `initialize` and an `attr_reader` to pull out stored values is needed, but I implemented no-ops for the rest of the Extractor interface to ensure we detect breakage in those due to this or future changes.

### CODE REVIEW

File: lib/graphql_metrics/extractor.rb [hunk 1]
```diff
@@ -93,6 +87,19 @@ def instrument(type, field)
       field.redefine { resolve(new_resolve_proc) }
     end
 
+    def extract_query_metrics(query)
```
Comment (chrisbutcher) [inline lib/graphql_metrics/extractor.rb]: `extract_metrics`, `extract` or `extract!` might be better names...

File: lib/graphql_metrics/extractor.rb [hunk 2]
```diff
@@ -106,6 +113,8 @@ def handle_extraction_exception(ex)
       raise ex
     end
 
+    private
```
Comment (chrisbutcher) [inline lib/graphql_metrics/extractor.rb]: :O Can't believe I was missing this.

File: lib/graphql_metrics/extractor.rb [hunk 3]
```diff
@@ -133,17 +142,19 @@ def extract_batch_loaders
...
 
+      duration = start_time && end_time ? end_time - start_time : nil
+
       query_extracted(
         {
           query_string: query.document.to_query_string,
           operation_type: query.selected_operation.operation_type,
           operation_name: query.selected_operation_name,
-          duration: end_time - start_time
+          duration: duration
```
Comment (chrisbutcher) [inline lib/graphql_metrics/extractor.rb]: Change in contract here. There's no sensible duration for a query that didn't run.

File: lib/graphql_metrics/extractor.rb [hunk 4]
```diff
@@ -164,7 +175,7 @@ def extract_field(irep_node)
           type_name: irep_node.owner_type.name,
           field_name: irep_node.definition.name,
           deprecated: irep_node.definition.deprecation_reason.present?,
-          resolver_times: ctx_namespace.dig(TIMING_CACHE_KEY, irep_node.ast_node) || [],
+          resolver_times: ctx_namespace&.dig(TIMING_CACHE_KEY, irep_node.ast_node) || [],
```
Comment (chrisbutcher) [inline lib/graphql_metrics/extractor.rb]: Likewise, `resolver_times` don't make sense to return if the extractor isn't run as instrumentation.

File: lib/graphql_metrics/extractor.rb [hunk 5]
```diff
@@ -93,6 +87,19 @@ def instrument(type, field)
       field.redefine { resolve(new_resolve_proc) }
     end
 
+    def extract!(query)
+      @query = query
+
+      extract_query
+
+      query.operations.each_value do |operation|
```
Comment (eapache) [inline lib/graphql_metrics/extractor.rb]: Thinking through something Scott mentioned... if `irep_selection` is already limited to the current operation, is it weird that we're extracting variables for *all* operations here?

Comment (chrisbutcher) [reply lib/graphql_metrics/extractor.rb]: 👀 I think you're right. I'll create an issue and some tests for this separately.

Comment (eapache) [reply lib/graphql_metrics/extractor.rb]: For the general metrics use case I think we wanted to log the use of fields in ops that weren't used, since they could still cause validation errors? So maybe it's the use of `irep_selection` that is incorrect?

File: lib/graphql_metrics/extractor.rb [hunk 6]
```diff
@@ -133,17 +142,19 @@ def extract_batch_loaders
     def extract_query
       return unless respond_to?(:query_extracted)
 
-      start_time = ctx_namespace[START_TIME_KEY]
-      return unless start_time
+      start_time = ctx_namespace[START_TIME_KEY] if ctx_namespace
```
Comment (swalkinshaw) [inline lib/graphql_metrics/extractor.rb]: maybe two separate returns?

```ruby
return unless ctx_namespace
return unless (start_time = ctx_namespace[START_TIME_KEY])
```

File: lib/graphql_metrics/extractor.rb (hunk 5 above)
Comment (swalkinshaw) [reply lib/graphql_metrics/extractor.rb]: We should probably ship this PR as is and deal with that separately anyway. It may even technically be a breaking change while this PR isn't.

File: lib/graphql_metrics/extractor.rb (hunk 6 above)
Comment (chrisbutcher) [reply lib/graphql_metrics/extractor.rb]: That would change the behaviour away from `return unless start_time || ctx_namespace.nil?`

I want to return early if the code is running as instrumentation, and we don't have a start_time, but don't want to return early when we're running `extract!` (and `ctx_namespace` isn't set). I may make this more explicit, and introduce a concept of runtime modes: one-off (extract!) vs. instrumentation, and check that instead of related values.

**Comment by chrisbutcher** [comment]
Updated to extract a new boolean via `variable_extracted`, `used_in_query` which indicates if the variable was actually used in the selected operation.

Also introduces a `running_as_instrumentation` mode, to more cleanly branch between code paths that only make sense when running as instrumentation, vs. an ad hoc extractor.

**Comment by chrisbutcher** [comment]
tldr: We / other clients can use this to conditionally log, or log differently unused variables.

**Comment by chrisbutcher** [comment]
(Note to self: Wrap the clock_gettime calls within the gem within something like GraphQLMetrics::TimedBatchExecutor.current_time, so that we can stub that, instead of the stdlib method, with `Process.stubs(:clock_gettime)`)

**Comment by chrisbutcher** [comment]
@swalkinshaw @eapache At your earliest convenience, please see the latest 2 commits (perhaps start with the changes to README.md in the first commit).

This introduces a breaking change (the meaning of `Extractor` changed), so I'll be bumping the gem to 2.0 after this PR.

```diff
-class LoggingExtractor < GraphQLMetrics::Extractor
+class LoggingExtractor < GraphQLMetrics::Instrumentation
```

File: README.md [hunk 7]
```diff
@@ -133,6 +134,30 @@ class LoggingExtractor < GraphQLMetrics::Extractor
 end
 ```
 
+You can also define ad hoc query Extractors that can work with instances of GraphQL::Query, for example:
```
Comment (swalkinshaw) [inline README.md]: Maybe mention these won't include resolver timings?

File: lib/graphql_metrics/timed_batch_executor.rb [hunk 8]
```diff
@@ -55,7 +55,7 @@ def serialize_loader_key(loader_key)
 
     def resolve(loader)
       @resolve_meta = {
-        start_time: Process.clock_gettime(Process::CLOCK_MONOTONIC),
+        start_time: Instrumentation.current_time,
```
Comment (swalkinshaw) [inline lib/graphql_metrics/timed_batch_executor.rb]: 🎉

File: lib/graphql_metrics/extractor.rb [hunk 9]
```diff
@@ -131,19 +72,21 @@ def extract_batch_loaders
     end
 
     def extract_query
-      return unless respond_to?(:query_extracted)
+      return unless query_extracted_method = extraction_method(:query_extracted)
 
-      start_time = ctx_namespace[START_TIME_KEY]
-      return unless start_time
+      start_time = instrumentor.ctx_namespace[Instrumentation::START_TIME_KEY] if instrumentor
```
Comment (swalkinshaw) [inline lib/graphql_metrics/extractor.rb]:
```suggestion
      if instrumentor
        start_time = instrumentor.ctx_namespace[Instrumentation::START_TIME_KEY]
        return unless start_time
      end
```

Is this the same?

File: lib/graphql_metrics/extractor.rb [hunk 10]
```diff
@@ -2,117 +2,58 @@
...
-
-    def before_query(query)
-      return unless extractor_defines_any_visitors?
-
-      ns = query.context.namespace(CONTEXT_NAMESPACE)
-      ns[TIMING_CACHE_KEY] = {}
-      ns[START_TIME_KEY] = Process.clock_gettime(Process::CLOCK_MONOTONIC)
-    rescue StandardError => ex
-      handle_extraction_exception(ex)
+    def initialize(instrumentor)
```
Comment (swalkinshaw) [inline lib/graphql_metrics/extractor.rb]: Doesn't this need 

```suggestion
    def initialize(instrumentor = nil)
```

It can't be instantiated without one right now. I guess it's meant to be overridden, but still seems like this is a better default to have? Not every extractor might have a need on its own to override `initialize`.

File: lib/graphql_metrics/extractor.rb (hunk 9 above)
Comment (chrisbutcher) [reply lib/graphql_metrics/extractor.rb]: That works in terms of readability, but then `start_time` is initialized within the `if/end` block scope, not the parent method.

This is both correct and easier to read: 
```ruby
start_time = if instrumentor
  instrumentor.ctx_namespace[Instrumentation::START_TIME_KEY].tap do |start_time|
    return unless start_time
  end
end
```

**Comment by chrisbutcher** [comment]
@swalkinshaw Here's where I settled on the instrumentor/start/end time checks in `extract_query`.

```ruby
start_time, end_time = if instrumentor
  start_time = instrumentor.ctx_namespace[Instrumentation::START_TIME_KEY]
  return unless start_time

  [start_time, Instrumentation.current_time]
end
```

File: lib/graphql_metrics/extractor.rb [hunk 11]
```diff
@@ -156,20 +101,24 @@ def extract_query
     end
 
     def extract_field(irep_node)
-      return unless respond_to?(:field_extracted)
+      return unless field_extracted_method = extraction_method(:field_extracted)
       return unless irep_node.definition
 
-      field_extracted(
+      resolver_times = if instrumentor
+        instrumentor.ctx_namespace.dig(Instrumentation::TIMING_CACHE_KEY, irep_node.ast_node)
```
Comment (eapache) [inline lib/graphql_metrics/extractor.rb]: I wonder if it would be cleaner to move logic like this into a method on the instrumentor, and then define a dummy instrumentor class to use instead of nil?

File: lib/graphql_metrics/extractor.rb [hunk 12]
```diff
@@ -2,117 +2,76 @@
...
 
-    def before_query(query)
-      return unless extractor_defines_any_visitors?
-
-      ns = query.context.namespace(CONTEXT_NAMESPACE)
-      ns[TIMING_CACHE_KEY] = {}
-      ns[START_TIME_KEY] = Process.clock_gettime(Process::CLOCK_MONOTONIC)
-    rescue StandardError => ex
-      handle_extraction_exception(ex)
+    def instrumentor
```
Comment (chrisbutcher) [inline lib/graphql_metrics/extractor.rb]: This is here since I'd rather default again at runtime if an extractor calls `instrumentor`, and the extractor overrode `initialize` but didn't call `super`.
